Add transform

symeneses opened the pull request:
It solves #8

symeneses commented on src/arviz_base/transform.py:
```diff
@@ -0,0 +1,68 @@
+    idata : DataTree-like
+        DataTree from which to extract the data.
+    transform_funcs : dict of {str : Callable}, optional
+        Dictionary with the functions to transform each variable, by default None
+        If ``None``, the values are taking from the group ``unconstrained_posterior`` .
```
Do you see the possibility to add some validations to restrict the kind of functions or the attributes (shape, dims, etc) of the transformed data?
I am thinking if we can raise errors if the output is not valid.

OriolAbril commented:
Let's try to start with a smallish scope function to take care of things we need now. If we later on see we need or people ask for general helpers for applying arbitrary functions to datasets we can consider a more global function or extending this one.

For now let's use `get_unconstrained_samples` as name, and restrict to functions that are `func(da: DataArray) -> DataArray`. We can then simplify the docstring of https://arviz-stats.readthedocs.io/en/latest/api/generated/arviz_stats.loo_moment_match.html for example as mentioned in https://github.com/arviz-devs/arviz-stats/issues/153 to one call to this new function followed by a 2nd call to `dataset_to_dataarray`.

symeneses commented on src/arviz_base/transform.py:
```diff
@@ -0,0 +1,68 @@
+                dims=da.dims,
+                coords=da.coords,
+                attrs=da.attrs,
+            )
```
To have the function signature as `func(da: DataArray) -> DataArray`, it's not possible to use `xr.apply_ufunc`.

symeneses commented:
> Needs to be a follow up in `arviz-stats` repo

Oops, yup, it can't be here 😅  

> I have tried using it a bit and my only issue was not having `var_names`+`filter_vars` in the function. When there are deterministics in `posterior` for example I don't want those propagated to `unconstrained_posterior`. Not completely sure `coords` will be very useful in this case, but it might be best to add it too for consistency. 

I added the `var_names`+`filter_vars`, about the coords, I am not sure how they should work, something like this:

```
coords : dict of {hashable: array_like}, optional
        Map of dimension names to coordinate values. Dimensions without coordinates
        values will keep their original values.
```

I added two tests using the function you suggested. As it's shown in the tests, if two variables have the same dimension and only one is transformed, there would be `nan` to fill the transformed variable. Should it be like this or it's better to validate that this doesn't happen?

OriolAbril commented on src/arviz_base/transform.py:
```diff
@@ -0,0 +1,100 @@
+    ----------
+    idata : DataTree-like
+        DataTree from which to extract the data.
+    var_names : str or list of str, optional
```
I think if we use the following we'll get a better type hint (I know we probably do have bad type hints in multiple places, but I think at least list and tuples are valid), but you know more about type hints so let me know if you think it should not be incorporated:

```suggestion
    var_names : str or sequence of str, optional
```

OriolAbril commented on tests/test_transform.py:
```diff
@@ -0,0 +1,182 @@
+    dt_uc = idata["unconstrained_posterior"]
+    # expected values of mu
+    transformed_mu = lagged_cumsum(dt_post["mu"])
+    # fill nan as dimension can't be removed
+    filling_mu = xr.DataArray(np.array([np.nan]), coords={"draw": [9]}, dims="draw")
+    expected_mu = xr.concat([transformed_mu, filling_mu], dim="draw")
```
I think this is ok, we can always change it if we end up fining a real case where this happens and is relevant. Code-wise we can simplify this using `pad`. I don't know why I can't add an actual suggestion and I haven't executed this code so some changes might be needed

```python
expected_mu = transformed_mu.pad(draw=(0, 1), constant_values=np.nan)
```

ref: https://docs.xarray.dev/en/latest/generated/xarray.DataArray.pad.html

symeneses replied:
`pad` is really what it's being done, I just had to assign the coords as they are fill with np.nan.

OriolAbril reviewed (commented):
> about the coords, I am not sure how they should work, something like this:

I started writing a comment on how to do it but then realized we might end up with strange behaviour because we might slice `ds_in` but not `ds_out` so we end up with nans everywhere after xarray aligns things. But there is also no need for dims or coords in `ds_in` and `ds_out` to be the same so then we might need two different coord arguments... Given I think we'll hardly ever need this (if ever) let's not add this for now, if someone complains they should have a clear usecase and we can adapt later on.

---

The two comments I left are very minimal, this is basically ready to merge.

OriolAbril commented:
The CI error has to do with numpyro, definitely unrelated to the PR but should be addressed. I will try to look into it at some point but it is not an area I am super comfortable with.
